libraries/plugins/es_objects/es_objects.cpp

```diff
@@ -0,0 +1,301 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <graphene/es_objects/es_objects.hpp>
+
+#include <fc/smart_ref_impl.hpp>
+
+#include <curl/curl.h>
+#include <graphene/chain/proposal_object.hpp>
+#include <graphene/chain/balance_object.hpp>
+#include <graphene/chain/market_object.hpp>
+
+#include <graphene/utilities/elasticsearch.hpp>
+
+
+namespace graphene { namespace es_objects {
+
+namespace detail
+{
+
+class es_objects_plugin_impl
+{
+   public:
+      es_objects_plugin_impl(es_objects_plugin& _plugin)
+         : _self( _plugin )
+      {  curl = curl_easy_init(); }
+      virtual ~es_objects_plugin_impl();
+
+      void updateDatabase( const vector<object_id_type>& ids , bool isNew);
+
+      graphene::chain::database& database()
+      {
+         return _self.database();
+      }
+
+      es_objects_plugin& _self;
+      std::string _es_objects_elasticsearch_url = "http://localhost:9200/";
+      uint32_t _es_objects_bulk_replay = 1000;
+      uint32_t _es_objects_bulk_sync = 2;
+      bool _es_objects_proposals = true;
+      bool _es_objects_accounts = true;
+      bool _es_objects_assets = true;
+      bool _es_objects_balances = true;
+      bool _es_objects_limit_orders = true;
+      bool _es_objects_logs = true;
+      CURL *curl; // curl handler
+      vector <std::string> bulk;
+      vector<std::string> prepare;
+   private:
+      void PrepareProposal(const proposal_object* proposal_object);
+      void PrepareAccount(const account_object* account_object);
+      void PrepareAsset(const asset_object* asset_object);
+      void PrepareBalance(const balance_object* balance_object);
+      void PrepareLimit(const limit_order_object* limit_object);
+};
+
+void es_objects_plugin_impl::updateDatabase( const vector<object_id_type>& ids , bool isNew)
+{
+
+   graphene::chain::database &db = database();
+
+   auto block_time = db.head_block_time();
+
+   // check if we are in replay or in sync and change number of bulk documents accordingly
+   uint32_t limit_documents = 0;
+   if((fc::time_point::now() - block_time) < fc::seconds(30))
```

Comment:
Is this the only way to figure out if we are in replay or normal mode?

Reply:
not that i know, this one seems to be effective as it is used in already implemented elasticsearch for account history plugin: https://github.com/bitshares/bitshares-core/blob/master/libraries/plugins/elasticsearch/elasticsearch_plugin.cpp#L198

of course open to ideas to change both.